keps/sig-auth/2579-psp-replacement/README.md

```diff
@@ -0,0 +1,969 @@
+```
+<prefix>/allow: <policy level>
+<prefix>/allow-version: <policy version>
+<prefix>/audit: <policy level>
+<prefix>/audit-version: <policy version>
+<prefix>/warn: <policy level>
+<prefix>/warn-version: <policy version>
+```
+
+**Allow:** Pods meeting the requirements of the allowed level are allowed. Violations are rejected
+in admission.
+
+**Audit:** Pods and [templated pods] meeting the requirements of the audit policy level are ignored.
+Violations are recorded in a `<prefix>/audit: <violation>` annotation on the
+audit event for the request. Audit annotations will **not** be applied to the pod objects
+themselves, as doing so would violate the non-mutating requirement.
+
+**Warn:** Pods and [templated pods] meeting the requirements of the warn level are ignored.
+Violations are returned in a user-facing warning message. Warn & audit modes are independent; if the
+functionality of both is desired, then both labels must be set.
+
+[templated pods]: #podtemplate-resources
+
+<<[UNRESOLVED]>>
+
+The mode key names are open to discussion. In particular, I'm not satisfied that `allow` and
+`audit/warn` carry slightly different meanings:
+
+- `allow` means "allow pods meeting this profile level & below" (i.e. deny above)
+- `audit` and `warn` mean "audit/warn on pods exceeding this profile level" (i.e. audit/warn above)
+
+~~Using `deny` for the enforcing level would be more consistent with audit/warn (deny above), but is
+a counter-intuitive user experience. One would expect `deny = privileged` to deny privileged pods,
+but it really means allow privileged pods (hence using `allow` for the key).~~ (Rejected)
+
+Another option is to change the `audit` and `warn` keys to something meaning "don't audit/warn at this
+level and bellow", but we haven't thought of a good name. The best alternative we've come up with is
+`audit-above` and `warn-above`, but that is cumbersome.
+
+A final option is to change `allow` to `enforce`, which is more ambiguous. The ambiguity makes the
+inconsistency with audit & warn less obvious, but doesn't help the user experience.
+
+<<[/UNRESOLVED]>>
+
+There are several reasons for controlling the policy directly through namespace labels, rather than
+through a separate object:
+
+- Using labels enables various workflows around policy management through kubectl, for example
+  issuing queries like `kubectl get namespaces -l
+  <prefix>/allow-version!=v1.22` to find namespaces where the enforcing
+  policy isn't pinned to the most recent version.
+- Keeping the options on namespaces allows atomic create-and-set-policy, as opposed to creating a
+  namespace and then creating a second object inside the namespace.
+- Policies settings are per-namespace singletons, and singleton objects are not well supported in
+  Kubernetes.
+- Labels are lighter-weight than namespace fields, making it clear that this is a policy layer on
+  top of namespaces, not inherent to namespaces itself.
+
```

Comment:
labels are indexed. I would think they would be heavier weight then non-searchable fields?

Reply:
Being able to be used with a label selector is one of the benefits and played a big part in the user experience aspect of locating/identifying namespaces using particular levels

Reply:
oh, this was in reference to the "Labels are lighter-weight than namespace fields" comment. You meant lighter-weight in terms of memory/performance, and the doc meant lighter-weight conceptually, since it would only be present if this integration was being used.